Tasks/Common/sdkutils.ts

```diff
@@ -1,6 +1,278 @@
+import { VSTSTaskManifest } from 'Common/vstsUtils'
+import fs = require('fs')
+import path = require('path')
+import { format, parse } from 'url'
+import tl = require('vsts-task-lib/task')
+
+export abstract class SdkUtils {
+    private static readonly agentTempDirectoryVariable: string = 'Agent.TempDirectory'
+    private static readonly userAgentPrefix: string = 'AWS-VSTS'
+    private static readonly userAgentSuffix: string = 'exec-env/VSTS'
+    private static readonly userAgentHeader: string = 'User-Agent'
+
+    // set on the integration test server so we validate the agent is set
+    // on all test builds that use sdk clients
+    private static readonly validateUserAgentEnvVariable: string = 'AWSVSTSTesting_ValidateUserAgent'
+
+    public static readResources(): void {
+        const taskManifestFile = path.join(__dirname, 'task.json')
+        tl.setResourcePath(taskManifestFile)
+        SdkUtils.setSdkUserAgentFromManifest(taskManifestFile)
+    }
+
+    public static readResourcesFromRelativePath(relativeResourcePath: string): void {
+        const taskManifestFile = path.join(__dirname, relativeResourcePath)
+        tl.setResourcePath(taskManifestFile)
+        SdkUtils.setSdkUserAgentFromManifest(taskManifestFile)
+    }
+
+    // Injects a custom user agent conveying extension version and task being run into the
+    // sdk so usage metrics can be tied to the tools.
+    public static setSdkUserAgentFromManifest(taskManifestFilePath: string): void {
+        if (fs.existsSync(taskManifestFilePath)) {
+            const taskManifest = JSON.parse(fs.readFileSync(taskManifestFilePath, 'utf8')) as VSTSTaskManifest
+            const version = taskManifest.version
+            const userAgentString = `${this.userAgentPrefix}/${version.Major}.${version.Minor}.${version.Patch} ${
+                    this.userAgentSuffix
+                }-${taskManifest.name}`
+                // tslint:disable-next-line:whitespace align
+            ;((AWS as any).util as any).userAgent = () => {
+                return userAgentString
+            }
+        } else {
+            console.warn(`Task manifest ${taskManifestFilePath} not found, cannot set custom user agent!`)
+        }
+    }
+
+    // prefer Agent.TempDirectory but if not available due to use of a lower agent version
+    // (it was added in agent v2.115.0), fallback to using TEMP
+    public static getTempLocation(): string {
+        let tempDirectory = tl.getVariable(this.agentTempDirectoryVariable)
+        if (!tempDirectory) {
+            tempDirectory = process.env.TEMP
+            console.log(`Agent.TempDirectory not available, falling back to TEMP location at ${tempDirectory}`)
+        }
+
+        return tempDirectory
+    }
+
+    // This next block has a huge amount of unsafe any. This is needed because it does a bunch of
+    // hacking around with the prototype. I'm scared to touch it so I will not.
```

Comment:
Consider creating an issue and putting the url in the comments here.

Reply:
I don't think it's an issue per say it's bad typescript but not bad JavaScript which os what this block really is